indexofknowledge/iok
# [ipfs-cyto] formatGraph and init traversal verification

## Motivation

This PR adds graph format verification (`formatGraph`), standardizes node ID calculation via CIDv1, and uses it to implement traversals.

### Graph format utils

Adds new functionality to ipfs-cytoscape package that:

* Use ipld-dag-cbor to calculate CID for CBOR serializable objects
* Wrap all that in utils/iok.sh script for manual use + for scraping. (see `iok.sh format_graph`)
* Update v4 to stop calculating sha256 hash directly but using new dag-cbor, sha2-256 IPLD standard

Bonus:

* Also fixes the logger, which was accidentally disabled in a previous commit.
* Disable lazy-loading for now since that's eating up performance. We can revisit this later on.
* fix v4 graph upload and download to new IoK graph format (was inconsistent previously)

### Traversal verification

Given a graph CID and a list of node IDs (traversedList) which we have traversed, output an IoK graph traversal state obj:
* grab the graph from IPFS using the CID.
* assert that all node IDs in traversedList are in the graph
* construct a traversalState, which looks like this for now:

```
{
    cid: <origin graph cid>,
    traversedNodes: <traversedList, but alpha sorted by b58 encoding>
}
```


## Test Plan

Added new unit tests for ipfs-cyto stuff. Also scraper should be using the updated funcs, all to be tested in CI.

Deploy in staging https://iok-rustie.firebaseapp.com/. 
* Check `bafyreihn52oh234xkz7dred2fg27vqhy2cj7csxd6c6fm36ra3o3qzw22e`
* Download JSON, verify format

## Related PRs or Issues

Resolves https://github.com/indexofknowledge/iok/issues/116 #96 and #94
